### PR TITLE
Sync ``libasr`` with LFortran

### PR DESCRIPTION
**Pull request by czgdp1807**
https://github.com/lfortran/lfortran/pull/2837

### CODE REVIEW

**Comment by certik** [comment]
The changes look good at first sight, but I have not carefully reviewed every single one of them.

If you have any doubts about some area, please point me to it.

Otherwise let's get tests to pass and then we'll do a final review.

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 1]
```diff
@@ -142,8 +142,8 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
         main_func = nullptr;
         avail_mem_loc = 0;
 
-        min_no_pages = 1000;  // fixed 64 Mb memory currently
-        max_no_pages = 1000;  // fixed 64 Mb memory currently
+        min_no_pages = 100;  // fixed 6.4 Mb memory currently
+        max_no_pages = 100;  // fixed 6.4 Mb memory currently
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:146]: To be reverted (TBR)

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 2]
```diff
@@ -160,7 +160,10 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
     }
 
     void import_function(ASR::Function_t* fn) {
-        if (ASRUtils::get_FunctionType(fn)->m_abi != ASR::abiType::BindJS) return;
+        if (ASRUtils::get_FunctionType(fn)->m_abi != ASR::abiType::BindC) return;
+        if (ASRUtils::get_FunctionType(fn)->m_deftype != ASR::deftypeType::Interface) return;
+        if (ASRUtils::get_FunctionType(fn)->m_abi != ASR::abiType::BindC) return;
+        if (ASRUtils::is_intrinsic_function2(fn)) return;
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:166]: TBR

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 3]
```diff
@@ -186,14 +189,6 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
...
                     }
                 }
-            } else if (ASR::is_a<ASR::Module_t>(*item.second)) {
-                ASR::Module_t *m = ASR::down_cast<ASR::Module_t>(item.second);
-                for (auto &item : m->m_symtab->get_scope()) {
-                    if (ASR::is_a<ASR::Function_t>(*item.second)) {
-                        ASR::Function_t *fn = ASR::down_cast<ASR::Function_t>(item.second);
-                        import_function(fn);
-                    }
-                }
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:196]: TBR

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 4]
```diff
@@ -1157,12 +1152,13 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
...
-
-        if (ASRUtils::get_FunctionType(x)->m_abi == ASR::abiType::BindC) {
-            // Skip C Intrinsic Functions
+        if (ASRUtils::get_FunctionType(x)->m_abi == ASR::abiType::BindC &&
+            ASRUtils::get_FunctionType(x)->m_deftype == ASR::deftypeType::Interface) {
+            if (ASRUtils::is_intrinsic_function2(&x)) {
+                diag.codegen_warning_label(
+                    "WASM: C Intrinsic Functions not yet supported",
+                    {x.base.base.loc}, std::string(x.m_name));
+            }
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:1161]: TBR

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 5]
```diff
@@ -1171,8 +1167,13 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
...
                     ASRUtils::symbol_get_past_external(sub_call.m_name));
                 if (ASRUtils::get_FunctionType(s)->m_abi == ASR::abiType::BindC &&
+                    ASRUtils::get_FunctionType(s)->m_deftype == ASR::deftypeType::Interface &&
                     ASRUtils::is_intrinsic_function2(s)) {
-                    // Skip functions that call into C Intrinsic Functions
+                    diag.codegen_warning_label(
+                        "WASM: Calls to C Intrinsic Functions are not yet "
+                        "supported",
+                        {x.m_body[i]->base.loc},
+                        "Function: calls " + std::string(s->m_name));
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:1176]: TBR

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 6]
```diff
@@ -2318,7 +2319,7 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
 
     void visit_LogicalConstant(const ASR::LogicalConstant_t &x) {
         bool val = x.m_value;
-        int a_kind = ASRUtils::extract_kind_from_ttype_t(x.m_type);
+        int a_kind = ((ASR::Logical_t *)(&(x.m_type->base)))->m_kind;
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:2322]: TBR

File: src/libasr/codegen/asr_to_wasm.cpp [hunk 7]
```diff
@@ -3216,15 +3205,12 @@ Result<Vec<uint8_t>> asr_to_wasm_bytes_stream(ASR::TranslationUnit_t &asr,
...
 
-    LCompilers::PassOptions pass_options;
-    pass_options.always_run = true;
-    pass_options.verbose = co.verbose;
-    pass_options.dump_all_passes = co.dump_all_passes;
+    co.po.always_run = true;
     std::vector<std::string> passes = {"pass_array_by_data", "array_op",
                 "implied_do_loops", "print_arr", "do_loops", "select_case",
-                "nested_vars", "unused_functions", "intrinsic_function"};
+                "intrinsic_function", "nested_vars", "unused_functions"};
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_wasm.cpp:3211]: TBR

File: src/libasr/diagnostics.cpp [hunk 8]
```diff
@@ -89,7 +89,7 @@ std::string Diagnostics::render(LocationManager &lm,
                 }
                 out += "\n\n";
                 out += bold + "Note" + reset
-                    + ": Please report unclear or confusing messages as bugs at\nhttps://github.com/lcompilers/lpython/issues.\n";
+                    + ": Please report unclear or confusing messages as bugs at\nhttps://github.com/lfortran/lfortran/issues.\n";
```
Comment (Shaikh-Ubaid) [inline src/libasr/diagnostics.cpp:92]: Should this be reverted?

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 9]
```diff
@@ -96,16 +100,6 @@ LFORTRAN_API void _lfortran_random_number(int n, double *v)
...
-LFORTRAN_API void _lfortran_init_random_seed(unsigned seed)
-{
-    srand(seed);
-}
-
-LFORTRAN_API void _lfortran_init_random_clock()
-{
-    srand((unsigned int)clock());
-}
-
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:108]: TBR

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 10]
```diff
@@ -1144,38 +1184,6 @@ LFORTRAN_API double_complex_t _lfortran_zatanh(double_complex_t x)
...
-LFORTRAN_API double _lfortran_dfix(double x)
-{
-    if (x > 0.0) {
-        return floor(x);
-    } else {
-        return ceil(x);
-    }
-}
-
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:1178]: TBR

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 11]
```diff
@@ -2021,13 +2094,7 @@ LFORTRAN_API void _lfortran_read_int64(int64_t *p, int32_t unit_num)
...
     if (unit_num == -1) {
         // Read from stdin
-        scanf(
-#ifdef HAVE_LFORTRAN_MACHO
-            "%lld"
-#else
-            "%ld"
-#endif
-            , p);
+        scanf("%lld", p);
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:2097]: TBR

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 12]
```diff
@@ -2041,13 +2108,7 @@ LFORTRAN_API void _lfortran_read_int64(int64_t *p, int32_t unit_num)
...
         fread(p, sizeof(*p), 1, filep);
     } else {
-        fscanf(filep,
-#ifdef HAVE_LFORTRAN_MACHO
-            "%lld"
-#else
-            "%ld"
-#endif
-            , p);
+        fscanf(filep, "%lld", p);
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.c:2111]: TBR

File: src/libasr/runtime/lfortran_intrinsics.h [hunk 13]
```diff
@@ -67,8 +67,6 @@ typedef double _Complex double_complex_t;
 
 LFORTRAN_API double _lfortran_sum(int n, double *v);
 LFORTRAN_API void _lfortran_random_number(int n, double *v);
-LFORTRAN_API void _lfortran_init_random_clock();
-LFORTRAN_API void _lfortran_init_random_seed(unsigned seed);
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.h:71]: TBR

File: src/libasr/runtime/lfortran_intrinsics.h [hunk 14]
```diff
@@ -170,10 +168,6 @@ LFORTRAN_API float _lfortran_satanh(float x);
 LFORTRAN_API double _lfortran_datanh(double x);
 LFORTRAN_API float_complex_t _lfortran_catanh(float_complex_t x);
 LFORTRAN_API double_complex_t _lfortran_zatanh(double_complex_t x);
-LFORTRAN_API float _lfortran_strunc(float x);
-LFORTRAN_API double _lfortran_dtrunc(double x);
-LFORTRAN_API float _lfortran_sfix(float x);
-LFORTRAN_API double _lfortran_dfix(double x);
```
Comment (Shaikh-Ubaid) [inline src/libasr/runtime/lfortran_intrinsics.h:176]: TBR

File: src/libasr/string_utils.cpp [hunk 15]
```diff
@@ -142,7 +158,6 @@ std::string str_escape_c(const std::string &s) {
             case '\n': o << "\\n"; break;
             case '\r': o << "\\r"; break;
             case '\t': o << "\\t"; break;
-            case '\v': o << "\\v"; break;
```
Comment (Shaikh-Ubaid) [inline src/libasr/string_utils.cpp:145]: TBR

File: src/bin/lpython.cpp [hunk 16]
```diff
@@ -241,12 +218,12 @@ int emit_asr(const std::string &infile,
...
-    pass_options.all_symbols_mangling = compiler_options.all_symbols_mangling;
-    pass_options.module_name_mangling = compiler_options.module_name_mangling;
-    pass_options.global_symbols_mangling = compiler_options.global_symbols_mangling;
-    pass_options.intrinsic_symbols_mangling = compiler_options.intrinsic_symbols_mangling;
+    pass_options.verbose = compiler_options.po.verbose;
+    pass_options.pass_cumulative = compiler_options.po.pass_cumulative;
+    pass_options.all_symbols_mangling = compiler_options.po.all_symbols_mangling;
+    pass_options.module_name_mangling = compiler_options.po.module_name_mangling;
+    pass_options.global_symbols_mangling = compiler_options.po.global_symbols_mangling;
+    pass_options.intrinsic_symbols_mangling = compiler_options.po.intrinsic_symbols_mangling;
```
Comment (Shaikh-Ubaid) [inline src/bin/lpython.cpp:226]: I think we need not create `pass_options` here. I think we can reuse the `compiler_options.po` here.

File: src/bin/lpython.cpp [hunk 17]
```diff
@@ -349,11 +326,11 @@ int emit_c(const std::string &infile,
...
-    pass_options.verbose = compiler_options.verbose;
-    pass_options.all_symbols_mangling = compiler_options.all_symbols_mangling;
-    pass_options.module_name_mangling = compiler_options.module_name_mangling;
-    pass_options.global_symbols_mangling = compiler_options.global_symbols_mangling;
-    pass_options.intrinsic_symbols_mangling = compiler_options.intrinsic_symbols_mangling;
+    pass_options.verbose = compiler_options.po.verbose;
+    pass_options.all_symbols_mangling = compiler_options.po.all_symbols_mangling;
+    pass_options.module_name_mangling = compiler_options.po.module_name_mangling;
+    pass_options.global_symbols_mangling = compiler_options.po.global_symbols_mangling;
+    pass_options.intrinsic_symbols_mangling = compiler_options.po.intrinsic_symbols_mangling;
```
Comment (Shaikh-Ubaid) [inline src/bin/lpython.cpp:333]: Similarly, I think we can reuse `compiler_options.po` here.

File: src/bin/lpython.cpp [hunk 18]
```diff
@@ -406,11 +383,11 @@ int emit_c_to_file(const std::string &infile, const std::string &outfile,
...
```
Comment (Shaikh-Ubaid) [inline src/bin/lpython.cpp:390]: Similarly

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 19]
```diff
@@ -2798,9 +2846,6 @@ PyMODINIT_FUNC PyInit_lpython_module_)" + fn_name + R"((void) {
             SET_INTRINSIC_NAME(Exp, "exp");
             SET_INTRINSIC_NAME(Exp2, "exp2");
             SET_INTRINSIC_NAME(Expm1, "expm1");
-            SET_INTRINSIC_NAME(Trunc, "trunc");
-            SET_INTRINSIC_NAME(Fix, "fix");
-            SET_INTRINSIC_NAME(FloorDiv, "floordiv");
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_c_cpp.h:2803]: TBR

File: src/libasr/codegen/asr_to_julia.cpp [hunk 20]
```diff
@@ -1899,8 +1899,6 @@ class ASRToJuliaVisitor : public ASR::BaseVisitor<ASRToJuliaVisitor>
             SET_INTRINSIC_NAME(Exp, "exp");
             SET_INTRINSIC_NAME(Exp2, "exp2");
             SET_INTRINSIC_NAME(Expm1, "expm1");
-            SET_INTRINSIC_NAME(Trunc, "trunc");
-            SET_INTRINSIC_NAME(Fix, "fix");
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_julia.cpp:1903]: TBR

File: src/libasr/pass/intrinsic_function_registry.h [hunk 21]
```diff
@@ -3263,10 +3482,6 @@ namespace IntrinsicScalarFunctionRegistry {
                    verify_function>>& intrinsic_function_by_id_db = {
         {static_cast<int64_t>(IntrinsicScalarFunctions::LogGamma),
             {&LogGamma::instantiate_LogGamma, &UnaryIntrinsicFunction::verify_args}},
-        {static_cast<int64_t>(IntrinsicScalarFunctions::Trunc),
-            {&Trunc::instantiate_Trunc, &UnaryIntrinsicFunction::verify_args}},
-        {static_cast<int64_t>(IntrinsicScalarFunctions::Fix),
-            {&Fix::instantiate_Fix, &UnaryIntrinsicFunction::verify_args}},
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:3269]: TBR

File: src/libasr/pass/intrinsic_function_registry.h [hunk 22]
```diff
@@ -3297,8 +3512,10 @@ namespace IntrinsicScalarFunctionRegistry {
             {&FMA::instantiate_FMA, &FMA::verify_args}},
         {static_cast<int64_t>(IntrinsicScalarFunctions::FlipSign),
             {&FlipSign::instantiate_FlipSign, &FlipSign::verify_args}},
-        {static_cast<int64_t>(IntrinsicScalarFunctions::FloorDiv),
-            {&FloorDiv::instantiate_FloorDiv, &FloorDiv::verify_args}},
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:3301]: TBR

File: src/libasr/pass/intrinsic_function_registry.h [hunk 23]
```diff
@@ -3363,30 +3582,12 @@ namespace IntrinsicScalarFunctionRegistry {
...
     };
 
     static const std::map<int64_t, std::string>& intrinsic_function_id_to_name = {
         {static_cast<int64_t>(IntrinsicScalarFunctions::LogGamma),
             "log_gamma"},
 
-        {static_cast<int64_t>(IntrinsicScalarFunctions::Trunc),
-            "trunc"},
-        {static_cast<int64_t>(IntrinsicScalarFunctions::Fix),
-            "fix"},
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:3389]: TBR

File: src/libasr/pass/intrinsic_function_registry.h [hunk 24]
```diff
@@ -3417,8 +3618,10 @@ namespace IntrinsicScalarFunctionRegistry {
             "fma"},
         {static_cast<int64_t>(IntrinsicScalarFunctions::FlipSign),
             "flipsign"},
-        {static_cast<int64_t>(IntrinsicScalarFunctions::FloorDiv),
-            "floordiv"},
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:3421]: TBR

File: src/libasr/pass/intrinsic_function_registry.h [hunk 25]
```diff
@@ -3481,29 +3686,13 @@ namespace IntrinsicScalarFunctionRegistry {
...
-            "SymbolicGetArgument"},
     };
 
 
     static const std::map<std::string,
         std::tuple<create_intrinsic_function,
                     eval_intrinsic_function>>& intrinsic_function_by_name_db = {
                 {"log_gamma", {&LogGamma::create_LogGamma, &LogGamma::eval_log_gamma}},
-                {"trunc", {&Trunc::create_Trunc, &Trunc::eval_Trunc}},
-                {"fix", {&Fix::create_Fix, &Fix::eval_Fix}},
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:3506]: TBR

File: src/libasr/pass/intrinsic_function_registry.h [hunk 26]
```diff
@@ -3519,7 +3708,8 @@ namespace IntrinsicScalarFunctionRegistry {
                 {"exp2", {&Exp2::create_Exp2, &Exp2::eval_Exp2}},
                 {"expm1", {&Expm1::create_Expm1, &Expm1::eval_Expm1}},
                 {"fma", {&FMA::create_FMA, &FMA::eval_FMA}},
-                {"floordiv", {&FloorDiv::create_FloorDiv, &FloorDiv::eval_FloorDiv}},
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:3522]: TBR

File: src/libasr/pass/intrinsic_function_registry.h [hunk 27]
```diff
@@ -3573,12 +3757,13 @@ namespace IntrinsicScalarFunctionRegistry {
                  id_ == IntrinsicScalarFunctions::Cos ||
                  id_ == IntrinsicScalarFunctions::Gamma ||
                  id_ == IntrinsicScalarFunctions::LogGamma ||
-                 id_ == IntrinsicScalarFunctions::Trunc ||
-                 id_ == IntrinsicScalarFunctions::Fix ||
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_function_registry.h:3577]: TBR

**Comment by Shaikh-Ubaid** [comment]
I shared some TBR (to be reverted) comments above. Also, I think there are changes related to Symbolic support that need to be reverted.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 28]
```diff
@@ -1032,9 +1082,14 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
         return free_fn;
     }
 
-    inline void call_lfortran_free_string(llvm::Function* fn) {
-        std::vector<llvm::Value*> args = {tmp};
-        builder->CreateCall(fn, args);
+    inline void call_lcompilers_free_strings() {
+        // if (strings_to_be_deallocated.n > 0) {
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_llvm.cpp:1086]: I think we need to uncomment this?

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 29]
```diff
@@ -205,14 +217,15 @@ void handle_float(char* format, double val, char** result) {
...
+    // TODO: This will work for up to `F65.60` but will fail for:
+    // print "(F67.62)", 1.23456789101112e-62_8
+    sprintf(dec_str, "%.*lf", (60-integer_length), decimal_part);
     memmove(dec_str,dec_str+2,strlen(dec_str));
 
     char* dot_pos = strchr(format, '.');
+    decimal_digits = atoi(++dot_pos);
     width = atoi(format + 1);
     if (dot_pos != NULL) {
-        dot_pos++;
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/runtime/lfortran_intrinsics.c:214]: @gptsarthak can you review the changes here related to the formatting.

**Review by Thirumalai-Shaktivel — APPROVED** [review]
Apart from the below concern, I think this PR LGTM!
Thank you!

File: src/libasr/runtime/lfortran_intrinsics.c (hunk 29 above)
Comment (gptsarthak) [reply src/libasr/runtime/lfortran_intrinsics.c:214]: I checked with a diffchecker and i dont see any difference between this and lfortran main, so it looks good to me.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 28 above)
Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:1086]: Actually this is an incorrect fix. Correct fix should work with both LFortran and LPython.

File: src/libasr/runtime/lfortran_intrinsics.c (hunk 29 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/runtime/lfortran_intrinsics.c:214]: Good, thanks for the review!

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 28 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_llvm.cpp:1086]: Okay, let's merge it as is, and then I will submit a new PR fixing them.

**Review by certik — APPROVED** [review]
I think this seems good to me. I don't see any blocker, hopefully I didn't miss anything.

**Review by Pranavchiku — APPROVED** [review]
Thanks for doing this! Looks good to me.

**Review by Shaikh-Ubaid — APPROVED** [review]
Thanks for this. It looks good to me!

**Pull request merged**